File: field-officer-app/modules/granite-ar/android/src/main/java/expo/modules/granitear/GraniteARModule.kt
package expo.modules.granitear

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class GraniteARModule : Module() {

  override fun definition() = ModuleDefinition {

    Name("GraniteAR")

    Events(
      "onChange"
    )

    View(GraniteARView::class) {

      Prop("tapToken") { view: GraniteARView, token: Int ->
        if (token > 0) {
          view.triggerCenterTap()
        }
      }

      Prop("resetToken") { view: GraniteARView, token: Int ->
        if (token > 0) {
          view.resetMeasurement()
        }
      }

      // Accelerometer motion-gate tuning, exposed so thresholds can be dialled
      // in on real hardware without rebuilding the native module.
      // [0] = moving threshold (m/s2), [1] = steady threshold (m/s2),
      // [2] = hold time (ms) the phone must stay calm before measuring resumes.
      Prop("motionSensitivity") { view: GraniteARView, values: FloatArray? ->
        if (values != null && values.size >= 3) {
          view.setMotionSensitivity(values[0], values[1], values[2].toInt())
        }
      }

      Events(
        "onStatus",
        "onPointSelected",
        "onOverlayUpdate",
        "onSteadyChange"
      )
    }
  }
}

File: field-officer-app/modules/granite-ar/ios/GraniteARModule.swift
import ExpoModulesCore

// Swift/ARKit counterpart to the Android GraniteARModule.kt. Exposes the same
// module name, props, and events so App.js can use GraniteARView without
// knowing which native platform backs it.
public class GraniteARModule: Module {
  public func definition() -> ModuleDefinition {
    Name("GraniteAR")

    View(GraniteARView.self) {
      Prop("tapToken") { (view: GraniteARView, token: Int) in
        view.tapToken = token
      }

      Prop("resetToken") { (view: GraniteARView, token: Int) in
        view.resetToken = token
      }

      Events("onStatus", "onPointSelected", "onOverlayUpdate")
    }
  }
}
